katcp/client.py

```diff
@@ -639,36 +642,41 @@ def unhandled_request(self, msg):
 
     @gen.coroutine
     def _install(self):
-        ioloop_before = self.ioloop
-        # Do stuff to put us on the IOLoop
-        self._logger.debug("Starting client loop for {0!r}"
-                           .format(self._bindaddr))
-        self.ioloop_thread_id = get_thread_ident()
-        self._tcp_client = tornado.tcpclient.TCPClient()
-        self._running.set()
-        yield self._connect()
-        if self._stream is None and not self._auto_reconnect:
-            raise KatcpClientError("Failed to connect to {0!r}"
-                                   .format(self._bindaddr))
+        self._stopped.clear()
         try:
-            yield self._connect_loop()
-        except Exception:
-            self._logger.exception('Unhandled exception in _connect_loop()')
-        finally:
+            ioloop_before = self.ioloop
+            # Do stuff to put us on the IOLoop
+            self._logger.debug("Starting client loop for {0!r}"
+                               .format(self._bindaddr))
+            self.ioloop_thread_id = get_thread_ident()
+            self._tcp_client = tornado.tcpclient.TCPClient()
+            self._running.set()
+            yield self._connect()
+            if self._stream is None and not self._auto_reconnect:
+                raise KatcpClientError("Failed to connect to {0!r}"
+                                       .format(self._bindaddr))
             try:
-                # Make sure everything is torn down properly
-                if ioloop_before == self.ioloop:
-                    # But only if we are still using the same ioloop. If the ioloop has
-                    # changed that means this client has been stopped and re-started
-                    # before we could get back to this finally clause. This would result
-                    # in us stopping the new ioloop. D'oh!
-                    self.ioloop.add_callback(self.stop)
-            except RuntimeError, e:
-                if str(e) == 'IOLoop is closing':
-                    # Seems the ioloop was stopped already, no worries.
-                    self._running.clear()
-            self._logger.info("Client connect loop for {0!r} finished."
-                              .format(self._bindaddr))
+                yield self._connect_loop()
+            except Exception:
+                self._logger.exception('Unhandled exception in _connect_loop()')
+            finally:
+                try:
+                    # Make sure everything is torn down properly
+                    if ioloop_before == self.ioloop:
+                        # But only if we are still using the same ioloop. If the ioloop has
+                        # changed that means this client has been stopped and re-started
+                        # before we could get back to this finally clause. This would result
+                        # in us stopping the new ioloop. D'oh!
```

Comment:
In that case (where we've restarted before getting here), we're still going to clear `_stopped` in the final `finally` clause. Is that a bug?

Reply:
Good question.  Think it is a bug.  If the user has changed the ioloop, but hasn't called `start` a second time then we would be OK.  However, if they have called `start` again, the behaviour would be incorrect.

Currently the `start` method fails if `_running` is already set.  If I include a check there for the new `_stopped` flag as well, then we should be safe.

Reply:
I tried to test this edge case, but doesn't seem like we could ever hit it:
- if the ioloop is unmanaged it cannot be changed to a new instance, even if we have stopped the client.  [This error](https://github.com/ska-sa/katcp-python/blob/232647424019c72c9b384e9bcfaad5f075009871/katcp/ioloop_manager.py#L59) is raised.
- if the ioloop is managed, stopping and starting again makes a new ioloop instance.  If you don't call `join` after `stop`, the next call to `start` [fails](https://github.com/ska-sa/katcp-python/blob/232647424019c72c9b384e9bcfaad5f075009871/katcp/ioloop_manager.py#L83) as the old ioloop's thread is still running.  If you do call `join`, the cleanup will wait for `_install` to be done, in which case we don't hit the problem. 
- if the ioloop was initially managed, we can change it to a new unmanaged ioloop instance, but only after calling `stop` and `join`.

Conclusion:  no need to add the check for `_stopped` in `start`.  I have added another test.

Reply:
Does that also mean that this `if` statement isn't needed?

Reply:
Yes.  Although an ill-advised user could modify `self.ioloop` directly.  It would be a little safer if it was changed to a read-only property.